### PR TITLE
Fix bug: when proxy resides in root namespace, the local destinationrule does not take effect

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
fixes :#13504

/assign @rshriram

### CODE REVIEW

**Comment by howardjohn** [comment]
I think the issue you linked that it fixes is wrong? That is a merged pr? Or is there no issue open for this bug?

File: pilot/pkg/model/push_context.go [hunk 1]
```diff
@@ -524,6 +524,18 @@ func (ps *PushContext) DestinationRule(proxy *Proxy, service *Service) *Config {
 		}
 	}
 
+	// when proxy config namespace equals root namespace
+	// search for the namespace local rules at last
+	if proxy.ConfigNamespace == ps.Env.Mesh.RootNamespace {
```
Comment (howardjohn) [inline pilot/pkg/model/push_context.go]: So this searches for rules marked with `exportTo: '.'` in the root namespace. I am not sure we want this? @rshriram would we expect these rules to be globally applicable in this case?

**Comment by howardjohn** [comment]
BTW we have a PR that adds tests to this https://github.com/istio/istio/pull/13482/files. The logic here is getting pretty complex so getting those in may be a good idea

File: pilot/pkg/model/push_context.go (hunk 1 above)
Comment (mbanikazemi) [reply pilot/pkg/model/push_context.go]: So with this change, we need to remove the following comment earlier in the code:
// NOTE: This does mean that we are effectively ignoring private dest rules in the config root namespace

Also is searching the public rules to be done before this (as is the case in the current PR) or after?

**Comment by mbanikazemi** [comment]
This PR is for Issue #13504

File: pilot/pkg/model/push_context.go (hunk 1 above)
Comment (rshriram) [reply pilot/pkg/model/push_context.go]: It should be after.

**Comment by hzxuzhonghu** [comment]
>I would suggest you remove the if condition on top of this function and change comment to indicate that proxies are not allowed in the config root namespace. We should add similar checks in the ads code to reject such proxies as well.

We can do this, but there is backward compatibility issue. For proxies in root-namespaces, if user upgrade istio, these proxies cannot work?

**Comment by hzxuzhonghu** [comment]
1.  So when proxy or gateway resides in root config namespace. I think the right order is 1. look for dest rule in local namespace(private-->public) 2. from service namespace (public)
The right select order I think is :

1. when the proxy or gateway resides in non root config namespace, the previous logic should remain.
   And the order is : 1. proxy namespace local 2. proxy namespace public 3. service namespace public 4. root config namespace public

**Pull request closed without merging**